src/services/Storage.tsx

```diff
@@ -0,0 +1,180 @@
+      return 'sentryTracking'
+    case 'selectedProfessions':
+      return 'selectedProfessions'
+    case 'isDevModeEnabled':
+      return 'devmode'
+    case 'progress':
+      return 'progress'
+    case 'cmsUrlOverwrite':
+      return 'cms'
+    case 'customDisciplines':
+      return 'customDisciplines'
+    case 'userVocabulary':
+      return 'userVocabulary'
+    case 'nextUserVocabularyId':
+      return 'userVocabularyNextId'
+    case 'favorites':
+      return 'favorites-2'
+  }
+}
+
+export const getStorageItemOr = async <T,>(key: string, defaultValue: T): Promise<T> => {
+  const value = await AsyncStorage.getItem(key)
+  return value ? JSON.parse(value) : defaultValue
+}
+
+export const getStorageItem = async <T extends keyof Storage>(key: T): Promise<Storage[T]> =>
+  getStorageItemOr(getStorageKey(key), defaultStorage[key])
+
+const setStorageItem = async <T extends keyof Storage>(key: T, value: Storage[T]): Promise<void> => {
+  await AsyncStorage.setItem(getStorageKey(key), JSON.stringify(value))
+}
+
+// https://github.com/react-native-async-storage/async-storage/issues/401#issuecomment-2508924008
+export class StorageCache {
+  private readonly listeners: Map<string, Set<() => void>> = new Map()
+  private readonly storage: Storage
+
+  private constructor(storage: Storage) {
+    this.storage = storage
+  }
+
+  static createForTesting = (): StorageCache => new StorageCache(newDefaultStorage())
+
+  static create = async (storage: Storage): Promise<StorageCache> => {
+    const storageCache = new StorageCache(storage)
+    await storageCache.migrate()
+    return storageCache
+  }
+
+  migrate = async (): Promise<void> => {
+    await migrateToNewFavoriteFormat(this)
+  }
+
+  /**
+   * Returns a storage item for the given key.
+   * The item should be treated as immutable and may not be modified.
+   *
+   * @param key The key of the storage item
+   */
+  getItem = <T extends keyof Storage>(key: T): Readonly<Storage[T]> => this.storage[key]
+
+  /**
+   * Returns a storage item that may be modified
+   *
+   * @param key The key of the storage item
+   */
+  getMutableItem = <T extends keyof Storage>(key: T): Storage[T] => JSON.parse(JSON.stringify(this.getItem(key)))
+
+  setItem = async <T extends keyof Storage>(key: T, value: Storage[T]): Promise<void> => {
+    this.storage[key] = value
+    await setStorageItem(key, value)
+    this.notifyListeners(key)
+  }
+
+  addListener = <T extends keyof Storage>(key: T, listener: () => void): (() => void) => {
+    if (!this.listeners.has(key)) {
+      this.listeners.set(key, new Set())
+    }
+    this.listeners.get(key)?.add(listener)
+
+    return () => {
+      this.listeners.get(key)?.delete(listener)
+    }
+  }
+
+  private notifyListeners = (key: keyof Storage) => {
+    this.listeners.get(key)?.forEach(listener => {
+      listener()
+    })
+  }
+}
+
+export const StorageCacheContext = createContext<StorageCache>(StorageCache.createForTesting())
```

Comment:
🙃 I think the `createForTesting` naming here is a little confusing as we use it for production as well. 
```suggestion
export const StorageCacheContext = createContext<StorageCache>(newDefaultStorage())
```

Reply:
I think `createDummy` kinda transports the same message, thats why I put `newDefaultStorage()` as my suggestion. But this is just a minor remark, so just pick what ever you feel like fits the most.

Reply:
Hmm, I think I did not fully understand your suggestion. Do you suggest that we should change the return type of `newDefaultStorage` to `StorageCache`? That would require making the constructor of `StorageCache` public, but in that case we could also just write `new StorageCache(newDefaultStorage())`. But I would like to keep the constructor private, to emphasize that there are only two ways of creating storage cache: One method that creates a dummy storage cache and one method that uses AsyncStorage.

My reasoning for`createDummy` is that it is alright because the "dummy" storage cache is only used as a default value and should never be actually accessed in production

Reply:
I missed that its private, sorry. Sounds good as described 👍

Reply:
Since this is not too important I will go ahead and merge the pr, just wanted to give my reasoning